src/components/application_manager/test/commands/mobile/on_hmi_status_notification_from_mobile_test.cc

```diff
@@ -0,0 +1,278 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "gtest/gtest.h"
+#include "application_manager/commands/mobile/on_hmi_status_notification_from_mobile.h"
+#include "commands/commands_test.h"
+#include <application_manager/smart_object_keys.h>
+#include "application_manager/message.h"
+#include "application_manager/application_manager.h"
+#include "application_manager/mock_message_helper.h"
+
+namespace test {
+namespace components {
+namespace commands_test {
+namespace mobile_commands_test {
+
+namespace {
+const uint32_t kConnectionKey = 1u;
+const connection_handler::DeviceHandle kHandle = 2u;
+}  // namespace
+
+namespace strings = application_manager::strings;
+
+using application_manager::commands::OnHMIStatusNotificationFromMobile;
+using application_manager::ProtocolVersion;
+using application_manager::ApplicationSet;
+using testing::Mock;
+using testing::Return;
+using testing::_;
+
+class OnHMIStatusNotificationFromMobileTest
+    : public CommandsTest<CommandsTestMocks::kIsNice> {
+ public:
+  MessageSharedPtr CreateMsgParams(
+      const mobile_apis::HMILevel::eType kHMIState) {
+    MessageSharedPtr msg = CreateMessage();
+    (*msg)[strings::params][strings::connection_key] = kConnectionKey;
+    (*msg)[strings::msg_params][strings::hmi_level] = kHMIState;
+    return msg;
+  }
+  ApplicationSet app_set_;
+  sync_primitives::Lock lock_;
+};
+
+TEST_F(OnHMIStatusNotificationFromMobileTest,
+       Run_CurrentStateForeground_SUCCESS) {
+  MessageSharedPtr msg = CreateMsgParams(mobile_apis::HMILevel::HMI_FULL);
+
+  SharedPtr<OnHMIStatusNotificationFromMobile> command =
+      CreateCommand<OnHMIStatusNotificationFromMobile>(msg);
+
+  MockAppPtr mock_app = CreateMockApp();
+  EXPECT_CALL(app_mngr_, application(kConnectionKey))
+      .WillOnce(Return(mock_app));
+  EXPECT_CALL(*mock_app, set_foreground(true));
+
+  EXPECT_CALL(*mock_app, device()).WillOnce(Return(kHandle));
+  EXPECT_CALL(app_mngr_, IsAppsQueriedFrom(kHandle)).WillOnce(Return(true));
+
+  DataAccessor<ApplicationSet> accessor(app_set_, lock_);
+  EXPECT_CALL(app_mngr_, applications()).WillOnce(Return(accessor));
+
+  EXPECT_CALL(*mock_app, protocol_version())
+      .WillRepeatedly(Return(ProtocolVersion::kV4));
+  EXPECT_CALL(*mock_app, is_foreground()).WillRepeatedly(Return(true));
+
+  command->Run();
```

Comment:
@pvvasilev, I think that in general unit tests can be improved by verifying call sequence - see:
https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#sequences
It is not complex and doesn't need much rework.